lib/bricks-net-api-types.js

```diff
@@ -0,0 +1,1176 @@
+	var _this = this;
+	
+	if (!(_this instanceof POSTFromFile)) {
+		_this = new POSTFromFile(url, file_name, content_type);
+		return _this;
+	}
+	
+	_this.file_name = "";
+	_this.content_type = "";
+	
+	cppArguments.assert('POSTFromFile', [
+		[
+			cppArguments.assertion('string', 'const std::string&', 'url'),
+			cppArguments.assertion('string', 'const std::string&', 'file_name'),
+			cppArguments.assertion('string', 'const std::string&', 'content_type'),
+			function (url, file_name, content_type) {
+				HTTPRequestBase.call(_this, url);
+	
+				_this.file_name = file_name;
+				_this.content_type = content_type;
+			}
+		]
+	], arguments);
+}
+inherits(POSTFromFile, HTTPRequestBase);
+
+
+function HTTPResponse() {
+	var _this = this;
+	
+	// The final URL. Will be equal to the original URL, unless redirects have been allowed and took place.
+	_this.url = "";
+	// HTTP response code.
+	_this.code = 0;
+	
+	promiseHelpers.makeThenable(_this);
+	promiseHelpers.makeThenableProperty(_this, 'url');
+	promiseHelpers.makeThenableProperty(_this, 'code');
+}
+
+function HTTPResponseWithBuffer() {
+	HTTPResponse.call(this);
+	
+	this.body = "";
+	
+	promiseHelpers.makeThenableProperty(this, 'body');
+}
+inherits(HTTPResponseWithBuffer, HTTPResponse);
+
+function HTTPResponseWithResultingFileName() {
+	HTTPResponse.call(this);
+	
+	// The file name into which the returned HTTP body has been saved.
+	this.body_file_name = "";
+	
+	promiseHelpers.makeThenableProperty(this, 'body_file_name');
+}
+inherits(HTTPResponseWithResultingFileName, HTTPResponse);
+
+
+// Response storage policy.
+// The default one is `KeepResponseInMemory()`, which can be omitted.
+// The alternative one is `SaveResponseToFile(file_name)`.
+function KeepResponseInMemory() {
+	if (!(this instanceof KeepResponseInMemory)) {
+		return new KeepResponseInMemory();
+	}
+}
+
+function SaveResponseToFile(file_name) {
+	if (!(this instanceof SaveResponseToFile)) {
+		return new SaveResponseToFile(file_name);
+	}
+	
+	var _this = this;
+	
+	_this.file_name = "";
+	
+	cppArguments.assert('SaveResponseToFile', [
+		[
+			cppArguments.assertion('string', 'const std::string&', 'file_name'),
+			function (file_name) {
+				_this.file_name = file_name;
+			}
+		]
+	], arguments);
+}
+
+
+function TemplatedHTTPRequestData(socket) {
+	var _this = this;
+	
+	_this._DEBUG_LOG = function () {};
+	//_this._DEBUG_LOG = function () { console.log.apply(console, arguments); };
+	
+	_this._socket = socket;
+	
+	_this.method_ = "";
+	_this.url_ = "";
+	_this.raw_path_ = "";
+	
+	_this.body_buffer_begin_ = null;
+	_this.body_buffer_end_ = null;
+	
+	_this.chunked_body_buffer_ = null;
+	
+	promiseHelpers.makeThenable(_this);
+	
+	// Read the response right in the constructor (like in the original code, but async).
+	(function () {
+		_this._DEBUG_LOG('TemplatedHTTPRequestData: Begin.');
+		
+		function strstr(buffer, str, start) {
+			//_this._DEBUG_LOG('TemplatedHTTPRequestData strstr:',
```

Comment:
The global conditional would be a nice feature to have, I thought of it before. But I decided to avoid it for now because I had to debug a single class while these debug statements are in multiple, so I'd have had to read through all these log entries to filter the lines I need. Usually this is done by a robust log routing system, e.g. winston, which I wanted to avoid either.